### PR TITLE
Add a toplevel directive to switch off debugging

### PR DESCRIPTION
**Pull request by garrigue**
At some time in the past, `Clflags.debug` has been set to true in the toplevel.
This makes very difficult to use `ocaml -dlambda` as debugging events get in the way.
This also disables some optimizations for code written in the toplevel.

This PR adds a direction `#debug` to change this flag.
Then one can disable debugging by writing
```ocaml
#debug false;;
```
and get some meaningful output for `-dlambda`

### CODE REVIEW

**Comment by gasche** [comment]
In trunk you should be able to use `ocaml -dlambda -dno-locations` which hides debug events.

(See #9208)

**Comment by garrigue** [comment]
I was not aware of this new flag :-)
Yet the two are not equivalent. `-dno-locations` only changes the way lambda code is dumped, but not the way it is generated, in particular it still disables standard lambda-code optimizations.
My motivating example was this one:
```ocaml
type _ t = A : int -> int t | B : bool -> bool t;;
class ['a] c (x : 'a t) =
  object method f =
    let f : type a. a t -> a = function A u -> u | B v -> v in f x
  end;;
```
With `ocaml -dlambda -dno-locations` one obtains:
```
(let (shared/140 =a [0: #"f"] shared/135 =a [0: #"f" #"*dummy method*"])
  (letrec
    (c/93
       (let
         (c_init/137 =
            (function class/117
              (let
                (x/95 =o
                   (apply (field 2 (global CamlinternalOO!)) class/117 #"")
                 ids/134 =
                   (apply (field 7 (global CamlinternalOO!)) class/117
                     shared/135)
                 f/102 =o (field 0 ids/134))
                (seq
                  (apply (field 9 (global CamlinternalOO!)) class/117 f/102
                    (function self-1/105
                      (let
                        (f/106 =
                           (function param/110
                             (switch* param/110
                              case tag 0:
                               (let (u/108 =a (field 0 param/110)) u/108)
                              case tag 1:
                               (let (v/109 =a (field 0 param/110)) v/109))))
                        (apply f/106 (field_computed self-1/105 x/95)))))
                  (function env/119 self/118 x/94
                    (let
                      (self/120 =
                         (apply (field 23 (global CamlinternalOO!)) self/118
                           class/117))
                      (seq (setfield_ptr_computed self/120 x/95 x/94)
                        self/120)))))))
         (apply (field 18 (global CamlinternalOO!)) shared/140 c_init/137)))
    (apply (field 1 (global Toploop!)) "c/93" c/93)))
```
but with `#debug false`, this becomes
```
(let (shared/135 =a [0: #"f" #"*dummy method*"])
  (letrec
    (c/93
       (let
         (c_init/137 =
            (function class/117
              (let
                (x/95 =o
                   (apply (field 2 (global CamlinternalOO!)) class/117 #"")
                 ids/134 =
                   (apply (field 7 (global CamlinternalOO!)) class/117
                     shared/135)
                 f/102 =o (field 0 ids/134))
                (seq
                  (apply (field 9 (global CamlinternalOO!)) class/117 f/102
                    (function self-1/105
                      (let (param/141 = (field_computed self-1/105 x/95))
                        (field 0 param/141))))
                  (function env/119 self/118 x/94
                    (let
                      (self/120 =
                         (apply (field 23 (global CamlinternalOO!)) self/118
                           class/117))
                      (seq (setfield_ptr_computed self/120 x/95 x/94)
                        self/120)))))))
         (apply (field 18 (global CamlinternalOO!)) [0: #"f"] c_init/137)))
    (apply (field 1 (global Toploop!)) "c/93" c/93)))
```
where the case analysis disappears.

**Comment by gasche** [comment]
I don't mind having directives to set the compilation state from the toplevel (maybe we could think of a more general approach than adding each flag one by one?), but I think that it's also worth thinking about what the use-cases are, and whether they would be better served by some other (complementary) features.

If I understand correctly, your use-case is that you use `ocaml -dlambda` to quickly/interactively explore the intermediate representation (IR) of various simple expressions. You have the problem that they may not be compiled in the same way as with the default batch compiler (which does not enable debug), so you would like to use `ocaml` *without* the Debug flag being set in this IR-inspection mode.

Wouldn't this be better served by a command-line flag that would do the opposite of `-g`, for example `-no-g`? If you decide in the command-line to look for `-dlambda` output, shouldn't it also be the convenient place to specify this?

(The reason we set debugging by default for `ocaml` is to get backtraces. Most uses of the toplevel are not performance-oriented, and backtraces are a huge usability boon for users. Arguably `-g` could be set by default for `ocamlc` as well.)

To summarize: I have nothing against `#debug b;;`, but I think that we could *also* add more convenient ways to do what you want to do.

**Comment by garrigue** [comment]
Having had to do lamba-level debugging recently (#10338), I still wish we had such an option.
Both this one and `-no-g` are fine with me.
A flag is more coherent with `-dlambda`, but the advantage of a directive is that it is possible to change it dynamically, when one wants to see both outputs for instance.

**Comment by gasche** [comment]
Yes, I think that having both a flag and a directive would be nice. Personally I wouldn't use the directive, but it's fine if it's part of your workflow, and not much more work.

**Review by gasche — APPROVED** [review]
I'm not sure I would use this interface myself, but if it is convenient to @garrigue there is no reason not to have it, it is a trivial, clean change. This is good to go after a rebase.
